### DISCUSSION TITLE
Indexing a document that has a dependency in multiple collections

### DISCUSSION THREAD

**Discussion by georgiosd** [ravendb/ravendb#18353] [General]
Hey folks

I've ran into a situation that I can't find a solution for.

I have a `Post` class that can have parents of several types, let's say `Forum` and `Company`. 
I have modelled pinned posts by storing an array in the parent, with information about the order of pinned posts.

I then want to index the posts and query for pinned posts alone:

```
public PostIndex()
{
    Map = posts => from post in posts
        let parent = (IHasPinnedPosts)LoadDocument<object>(post.ParentId)
        select new Result
        {
            IsPinned = parent.PinnedPosts.Any(x => x.PostId == post.Id),
        };
}
```

This compiles, but fails to set `IsPinned` properly, because the collection name is missing on the `Load` call. 

I worked around it by copying the `IsPinned` property to the post itself.

Is there another solution that doesn't require that?

**Comment by ayende** [comment]
You have to specify which collections to look at explicitly, like so:

```
        let parent = (IHasPinnedPosts)(LoadDocument<Forum>(post.ParentId)  ?? LoadDocument<Company>(post.ParentId))

```

**Reply by georgiosd** [reply]
Would it make sense to make a way to dynamically specify the collection? I tried deriving it from the ID but the LINQ was not supported.

**Reply by ayende** [reply]
You must be able to provide the collection to the `Load` call, before you load it.

**Reply by georgiosd** [reply]
Yes, I can do that, but it has to be C# code whether inline or as an additional source, but from my tests that was not supported.

**Reply by ayende** [reply]
Note that this is meant to be something you do in the C# client index definition.
For the serve side, you'll use:


```
        let parent = LoadDocument("Forums", post.ParentId)  ?? LoadDocument("Companies", post.ParentId)

```

**Reply by georgiosd** [reply]
What I meant is, it would be nice to be able to `LoadDocument(id, id.Split('/').First())` either directly in the definition or in a helper inside additional sources. Probably low yield if no one else asked for this in six versions :)

**Reply by ayende** [reply]
You _can_ do that, this works:


```
from e in docs.Employees
where e.ReportsTo != null
let m = LoadDocument(e.ReportsTo, e.ReportsTo.Split("/")[0])
select new { e.FirstName, ManagerName = m.FirstName }
```

**Reply by georgiosd** [reply]
Sadly it doesn't.

This:
```
let parent = (IHasPinnedPosts)LoadDocument<object>(post.ParentId, post.ParentId.Split("/").First())
```
Doesn't compile at all, with error:
```
error CS0854: An expression tree may not contain a call or invocation that uses optional arguments
```

This doesn't work:
```
let parent = (IHasPinnedPosts)LoadDocument<object>(post.ParentId, post.ParentId.Split('/', StringSplitOptions.None).First())
```
with exception:
```
IndexCompilationException: Failed to create index PostIndex, IndexDefinitionProperty='', ProblematicText='' ---> System.InvalidOperationException: Invalid argument in LoadDocument. String constant was expected but was 'Call' with value 'post.ParentId.Split("/", None).First()'.
```
Also tried this, with similar error
```
let parent = (IHasPinnedPosts)LoadDocument<object>(post.ParentId, post.ParentId.Split(new char[] { '/' })[0])
```
Note: .NET 6.0

**Reply by ayende** [reply]
Please note, there is a distinction between the C# code on your end and what the server side can accept.
If you define the index as a string, that will work as expected.
The issue is that the value is generic method on the client side, not on the server side (we convert it during client side processing)

**Comment by grisha-kotler** [comment]
The way to do it is to use:
```
let parent = LoadDocument(post.ParentId, "Forums")  ?? LoadDocument(post.ParentId, "Companies")
```

We might support dynamic collection name in `LoadDocument`:
https://issues.hibernatingrhinos.com/issue/RavenDB-22342/Support-LoadDocument-with-dynamic-collection-name
